akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala

```diff
@@ -36,240 +38,214 @@ private object PoolSlot {
-          val actor = system.actorOf(
-            Props(new SlotProcessor(slotIx, connectionFlow)).withDeploy(Deploy.local),
-            name)
-          ActorProcessor[SlotCommand, List[ProcessorOut]](actor)
-        }.mapConcat(ConstantFun.scalaIdentityFunction)
-      }
-      val split = b.add(Broadcast[ProcessorOut](2))
-
-      slotProcessor ~> split.in
-
-      new FanOutShape2(
-        slotProcessor.in,
-        split.out(0).collect { case ResponseDelivery(r) ⇒ r }.outlet,
-        split.out(1).collect { case r: RawSlotEvent ⇒ r }.outlet)
+      val log = ActorMaterializerHelper.downcast(m).logger
+      val slotProcessor: FlowShape[SlotCommand, ProcessorOut] = b.add(new SlotProcessor(slotIx, connectionFlow, log))
+      val partition = b.add(Partition[ProcessorOut](2, {
+        case ResponseDelivery(r) ⇒ 0
+        case r: RawSlotEvent     ⇒ 1
+      }))
+      slotProcessor.out ~> partition.in
+
+      val deliveries = partition.out(0).map(_.asInstanceOf[ResponseDelivery].response)
+      val events = partition.out(1).map(_.asInstanceOf[RawSlotEvent])
+      new FanOutShape2[SlotCommand, ResponseContext, RawSlotEvent](slotProcessor.in, deliveries.outlet, events.outlet)
     }
 
-  import ActorPublisherMessage._
-  import ActorSubscriberMessage._
-
   /**
    * An actor managing a series of materializations of the given `connectionFlow`.
-   * To the outside it provides a stable flow stage, consuming `SlotCommand` instances on its
-   * input (ActorSubscriber) side and producing `List[ProcessorOut]` instances on its output
-   * (ActorPublisher) side.
+   *
+   * To the outside it provides a stable flow stage, consuming `SlotCommand` instances on its internal connection flows.
+   *
    * The given `connectionFlow` is materialized into a running flow whenever required.
    * Completion and errors from the connection are not surfaced to the outside (unless we are
    * shutting down completely).
    */
-  private class SlotProcessor(slotIx: Int, connectionFlow: Flow[HttpRequest, HttpResponse, Any])(implicit fm: Materializer)
-    extends ActorSubscriber with ActorPublisher[List[ProcessorOut]] with ActorLogging {
-    var exposedPublisher: akka.stream.impl.ActorPublisher[Any] = _
-    var inflightRequests = immutable.Queue.empty[RequestContext]
+  private class SlotProcessor(slotIx: Int, connectionFlow: Flow[HttpRequest, HttpResponse, Any], log: LoggingAdapter)(implicit fm: Materializer)
+    extends GraphStage[FlowShape[SlotCommand, ProcessorOut]] {
+
+    val in: Inlet[SlotCommand] = Inlet("SlotProcessor.in")
+    val out: Outlet[ProcessorOut] = Outlet("SlotProcessor.out")
+
+    override def shape: FlowShape[SlotCommand, ProcessorOut] = new FlowShape(in, out)
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) 
+      with InHandler with OutHandler { self ⇒
+      private final val inflightRequests = new java.util.ArrayDeque[RequestContext]()
```

Comment:
I wonder if we could do without it...
It's those requests which we have sent into the connection flow, but have no response for yet.
So if connection terminates on request 1, we should reconnect and send 2 on a new connection